File: builtin/http.js
var tm = process.binding('tm');
var http_parser = process.binding('http_parser');

var url = require('url');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var net = require('net');


/**
 * ServerResponse
 */

function ServerResponse (req, socket) {
  this.req = req;
  this.socket = socket;
  this.headers = {};
  this._header = false;
  this._closed = false;

  var self = this;
  this.req.on('close', function () {
    self._closed = true;
  })
}

// TODO stream
util.inherits(ServerResponse, EventEmitter);

ServerResponse.prototype.setHeader = function (name, value) {
  if (this._header) {
    throw "Already wrote HEAD";
  }
  this.headers[String(name).toLowerCase()] = value;
};

ServerResponse.prototype.setHeaders = function (headers) {
  for (var key in headers) {
    this.setHeader(key, headers[key]);
  }
};

ServerResponse.prototype.writeHead = function (status, headers) {
  if (this._header) {
    throw "Already wrote HEAD";
  }

  if (headers) {
    this.setHeaders(headers);
  }
  this.socket.write('HTTP/1.1 ' + status + ' OK\r\n');
  for (var key in this.headers) {
    this.socket.write(key + ': ' + this.headers[key] + '\r\n');
  }
  this.socket.write('Transfer-Encoding: chunked\r\n');
  this.socket.write('\r\n');
  this._header = true;
  // console.log('writing headers', headers);
};

ServerResponse.prototype.write = function (data) {
  if (!this._header) {
    this.writeHead(200);
  }

  this.socket.write(Number(data.length).toString(16));
  this.socket.write('\r\n');
  this.socket.write(data);
  this.socket.write('\r\n');
}

ServerResponse.prototype.end = function (data) {
  if (!this._header) {
    this.writeHead(200);
  }

  // TODO
  if (data != null) {
    this.write(data);
  }
  this._closed = true;
  this.socket.write('0\r\n\r\n');
  this.socket.close();
};


/**
 * ServerRequest
 */

function ServerRequest (socket) {
  var self = this;

  this.headers = {};
  this.socket = socket;
  this.url = null;

  function js_wrap_function (fn) {
    return function () {
      return fn.apply(null, [this].concat(arguments));
    }
  }

  var lastheader;
  var parser = http_parser.new('request', {
    onMessageBegin: js_wrap_function(function () {
    }),
    onUrl: js_wrap_function(function (url) {
      self.url = url;
    }),
    onHeaderField: js_wrap_function(function (field) {
      lastheader = field;
    }),
    onHeaderValue: js_wrap_function(function (value) {
      self.headers[lastheader.toLowerCase()] = value;
    }),
    onHeadersComplete: js_wrap_function(function (method) {
      self.method = method;
      self.emit('request', self);
    }),
    onBody: js_wrap_function(function (body) {
      self.emit('data', body);
    }),
    onMessageComplete: js_wrap_function(function () {
      self.emit('close');
      self._closed = true;
      // TODO close
    }),
    onError: js_wrap_function(function (err) {
      self.socket.emit('error', err)
    })
  })
  this.socket.on('data', function (data) {
    // console.log('received', data.length, data.substr(0, 15));
    parser.execute(data, 0, data.length);
  })
}

util.inherits(ServerRequest, EventEmitter);




function HTTPServer () {
  var self = this;
  this.socket = net.createServer(function (socket) {
    var request = new ServerRequest(socket);
    var response = new ServerResponse(request, socket);
    request.on('request', function () {
      self.emit('request', request, response)
    });
  });
}

util.inherits(HTTPServer, EventEmitter);

HTTPServer.prototype.listen = function (port, ip) {
  this.socket.listen(port, ip);
};


/**
 * HTTPIncomingResponse
 */

function HTTPIncomingResponse (data) {
  this.headers = {};
}

// TODO stream
util.inherits(HTTPIncomingResponse, EventEmitter);

HTTPIncomingResponse.prototype.setEncoding = function () {
  // TODO
};


/**
 * HTTPOutgoingRequest
 */

function HTTPOutgoingRequest (port, host, path, method) {
  if (!host.match(/^[0-9.]+$/)) {
    var ipl = tm.hostname_lookup(host);
    if (ipl == 0) {
      throw new Error('Could not lookup hostname.');
    }
    var ip = [(ipl >> 0) & 0xFF, (ipl >> 8) & 0xFF, (ipl >> 16) & 0xFF, (ipl >> 24) & 0xFF].join('.');
  } else {
    var ip = host;
  }

  if (path[0] != '/') {
    path = '/' + path;
  }

  var self = this;
  this.socket = net.connect(port, ip, function () {
    self.socket.write(method + ' ' + path + ' HTTP/1.1\r\nHost: ' + host + '\r\n\r\n');
    // self.emit('connect');
  })

  function js_wrap_function (fn) {
    return function () {
      return fn.apply(null, [this].concat(arguments));
    }
  }

  var response, lastheader;
  var parser = http_parser.new('response', {
    onMessageBegin: js_wrap_function(function () {
      response = new HTTPIncomingResponse();
    }),
    onUrl: js_wrap_function(function (url) {
      // nop
    }),
    onHeaderField: js_wrap_function(function (field) {
      lastheader = field;
    }),
    onHeaderValue: js_wrap_function(function (value) {
      response.headers[lastheader.toLowerCase()] = value;
    }),
    onHeadersComplete: js_wrap_function(function (obj) {
      response.statusCode = obj.status_code
      self.emit('response', response);
    }),
    onBody: js_wrap_function(function (body) {
      response.emit('data', body);
    }),
    onMessageComplete: js_wrap_function(function () {
      response.emit('close');
      // TODO close
      self.socket.close();
    })
  })
  this.socket.on('data', function (data) {
    var nparsed = parser.execute(data, 0, data.length);
    if (nparsed != data.length) {
      this.socket.emit('error', 'Could not parse tokens at character #' + String(nparsed));
    }
  })
}

util.inherits(HTTPOutgoingRequest, EventEmitter);

HTTPOutgoingRequest.prototype.end = function () {
  // TODO
  // console.log('end');
};


/**
 * Public API
 */

exports.request = function (opts, onresponse) {
  var req = new HTTPOutgoingRequest(opts.port || 80, opts.hostname, opts.path || '', opts.method || 'GET');
  onresponse && req.on('response', onresponse);
  return req;
};

exports.get = function (opts, onresponse) {
  if (typeof opts == 'string') {
    opts = url.parse(opts);
  }
  opts.method = 'GET';

  var req = exports.request(opts, onresponse);
  req.end();
  return req;
};

exports.createServer = function (onrequest) {
  var server = new HTTPServer();
  onrequest && server.on('request', onrequest);
  return server;
};

exports.ServerResponse = ServerResponse;
exports.ServerRequest = ServerRequest;

File: builtin/http.lua
function (_ENV)local string, math, print, type, pairs = nil, nil, nil, nil, nil;local _module = _obj({exports=_obj({})}); local exports, module = _module.exports, _module;




























local  util,  EventEmitter,  net,  ServerResponse,  ServerRequest,  HTTPServer,  HTTPIncomingResponse,  HTTPOutgoingRequest =  util,  EventEmitter,  net,  ServerResponse,  ServerRequest,  HTTPServer,  HTTPIncomingResponse,  HTTPOutgoingRequest;

 ServerResponse = (function (this, req,  socket)  local  self =  self;
  (this).req =  req;
  (this).socket =  socket;
  (this).headers =  _obj({  });
  (this)._header =  (false);
  (this)._closed =  (false);

   self = this;if 
  (this).req:on(("close"),  (function (this)  
(
    self)._closed =  (true);end))
 then end; end);

 ServerRequest = (function (this, socket)  local  self,  lastheader,  parser =  self,  lastheader,  parser;
   self = this;

  (this).headers =  _obj({  });
  (this).socket =  socket;
  (this).url =  (null);

   lastheader = nil;
  
 parser =  _new( tm__http__parser, ("request"),  








_obj({  ["on_message_begin"]= (function (this)  
end),  ["on_url"]= (function (this, url)  
(
      self).url =  url;end),  ["on_header_field"]= (function (this, field)  

      lastheader =  field;end),  ["on_header_value"]= (function (this, value)  
((
      self).headers)[lastheader:toLowerCase()] =  value;end),  ["on_headers_complete"]= (function (this, method)  
(
      self).method =  method;if 
      self:emit(("request"),  self) then end; end),  ["on_body"]= (function (this, body)  
if 
      self:emit(("data"),  body) then end; end),  ["on_message_complete"]= (function (this)  

if 
      self:emit(("close")) then end; (
      self)._closed =  (true);end),  ["on_error"]= (function (this, err)  if (
      self).socket:emit(("error"),  err)
     then end; end)}));if (this).socket:on(("data"),  (function (this, data)  

if 
    parser:write(data) then end; end))
 then end; end);




 HTTPServer = (function (this)  
local  self =  self;
   self = this;
  (this).socket =  net:createServer((function (this, socket)  
local  request,  response =  request,  response;
     request =  _new( ServerRequest, socket);
     response =  _new( ServerResponse, request,  socket);if 
    request:on(("request"),  (function (this)  if 
      self:emit(("request"),  request,  response)
     then end; end)) then end; end));end);

 HTTPIncomingResponse = (function (this, data)  

  (this).headers =  _obj({  });end);

 HTTPOutgoingRequest = (function (this, port,  host,  path,  method)  local  ipl,  ip,  self,  response,  lastheader,  parser =  ipl,  ip,  self,  response,  lastheader,  parser;
   ipl =  tm__hostname__lookup(global, host);
  if _truthy((ipl == (0))) then 

    _error( _new( Error, ("Could not lookup hostname.")))end
   ip =  _arr({[0]=_bit.band(_bit.rshift(ipl,  (0)),  (255)),  _bit.band(_bit.rshift(ipl,  (8)),  (255)),  _bit.band(_bit.rshift(ipl,  (16)),  (255)),  _bit.band(_bit.rshift(ipl,  (24)),  (255))}):join(("."));

   self = this;
  (this).socket =  net:connect(port,  ip,  (function (this)  

if (
    self).socket:write((((((method + (" /")) + path) + (" HTTP/1.1\r\nHost: ")) + host) + ("\r\n\r\n"))) then end; end))

  ; response = nil; lastheader = nil;
  
 parser =  _new( tm__http__parser, ("response"),  








_obj({  ["on_message_begin"]= (function (this)  

      response =  _new( HTTPIncomingResponse);end),  ["on_url"]= (function (this, url)  end),  ["on_header_field"]= (function (this, field)  

      lastheader =  field;end),  ["on_header_value"]= (function (this, value)  
((
      response).headers)[lastheader:toLowerCase()] =  value;end),  ["on_headers_complete"]= (function (this)  
if 
      console:log(response) then end; if 
      self:emit(("response"),  response) then end; end),  ["on_body"]= (function (this, body)  
if 
      response:emit(("data"),  body) then end; end),  ["on_message_complete"]= (function (this)  

if 
      response:emit(("close")) then end; if (
      self).socket:close() then end; end),  ["on_error"]= (function (this, err)  if (
      self).socket:emit(("error"),  err)
     then end; end)}));if (this).socket:on(("data"),  (function (this, data)  
if 
    parser:write(data) then end; end))
 then end; end); util =  require(global, ("util"));
 EventEmitter = ( require(global, ("events"))).EventEmitter;
 net =  require(global, ("net"));if 
util:inherits(ServerResponse,  EventEmitter) then end; ((

ServerResponse).prototype).setHeader =  (function (this, name,  value)  

  if _truthy((this)._header) then 

    _error( ("Already wrote HEAD"))end(
  (this).headers)[String(global, name):toLowerCase()] =  value;end);((

ServerResponse).prototype).setHeaders =  (function (this, headers)  
local  key =  key;
  for key in _pairs( headers) do 
if 
    this:setHeader(key,  headers[key]) then end; endend);((

ServerResponse).prototype).writeHead =  (function (this, status,  headers)  

local  key =  key;
  if _truthy((this)._header) then 

    _error( ("Already wrote HEAD"))end

  if _truthy(headers) then 
if 
    this:setHeaders(headers) then end; endif 
  (this).socket:write(((("HTTP/1.1 ") + status) + (" OK\r\n"))) then end; 
  for key in _pairs( (this).headers) do 
if 
    (this).socket:write((((key + (": ")) +( (this).headers)[key]) + ("\r\n"))) then end; endif 
  (this).socket:write(("Transfer-Encoding: chunked\r\n")) then end; if 
  (this).socket:write(("\r\n")) then end; 
  (this)._header =  (true);end);((

ServerResponse).prototype).write =  (function (this, data)  

  if (not _truthy((this)._header)) then 
if 
    this:writeHead((200)) then end; endif 

  (this).socket:write(Number(global, data.length):toString((16))) then end; if 
  (this).socket:write(("\r\n")) then end; if 
  (this).socket:write(data) then end; if 
  (this).socket:write(("\r\n")) then end; end)

;((ServerResponse).prototype)["end"] =  (function (this, data)  



  if (not _truthy((this)._header)) then 
if 
    this:writeHead((200)) then end; end
  if (data ~=  (null)) then 
if 
    this:write(data) then end; end
  (this)._closed =  (true);if 
  (this).socket:write(("0\r\n\r\n")) then end; if 
  (this).socket:close() then end; end);if 

util:inherits(ServerRequest,  EventEmitter) then end; if 

util:inherits(HTTPServer,  EventEmitter) then end; ((

HTTPServer).prototype).listen =  (function (this, port,  ip)  
if 
  (this).socket:listen(port,  ip) then end; end);if 
util:inherits(HTTPIncomingResponse,  EventEmitter) then end; ((

HTTPIncomingResponse).prototype).setEncoding =  (function (this)  

end);if 

util:inherits(HTTPOutgoingRequest,  EventEmitter) then end; ((

HTTPOutgoingRequest).prototype)["end"] =  (function (this)  


end);(

exports).request =  (function (this, opts,  onresponse)  
local  req =  req;
   req =  _new( HTTPOutgoingRequest, (opts).port or  (80), ( opts).hostname, ( opts).path or  (""), ( opts).method or  ("GET"));if 
  onresponse and  req:on(("response"),  onresponse) then end; 
  if true then return  req; end;end);(

exports).createServer =  (function (this, onrequest)  
local  server =  server;
   server =  _new( HTTPServer);if 
  onrequest and  server:on(("request"),  onrequest) then end; 
  if true then return  server; end;end);(

exports).ServerResponse =  ServerResponse;(
exports).ServerRequest =  ServerRequest;return _module.exports;end 